File: examples/c/compile.c
/* A runnable C example: compile a strategy spec (dry run) through the
 * wickra-compile C ABI and print the raw JSON manifest. Every language example
 * uses the same spec and prints the same project_hash. */
#include <stdio.h>
#include <stdlib.h>

#include "wickra_compile.h"

static const char *CMD =
    "{\"cmd\":\"compile\",\"dry_run\":true,\"spec\":"
    "{\"strategy\":{\"symbol\":\"btcusdt\",\"timeframe\":\"1h\","
    "\"indicators\":{\"fast\":{\"type\":\"Sma\",\"params\":[10]},"
    "\"slow\":{\"type\":\"Sma\",\"params\":[30]}},"
    "\"entry\":{\"cross_above\":[\"fast\",\"slow\"]},"
    "\"exit\":{\"cross_below\":[\"fast\",\"slow\"]},"
    "\"sizing\":{\"type\":\"fixed_qty\",\"qty\":1}},"
    "\"target\":{\"kind\":\"wasm\"},\"crate_name\":\"demo\"}}";

int main(void) {
    WickraCompiler *compiler = wickra_compile_new();
    if (!compiler) {
        fprintf(stderr, "failed to build compiler\n");
        return 1;
    }

    /* Length-out protocol: learn the length, then read into a caller buffer. */
    int len = wickra_compile_command(compiler, CMD, NULL, 0);
    if (len < 0) {
        fprintf(stderr, "command failed: code %d\n", len);
        wickra_compile_free(compiler);
        return 1;
    }
    char *buf = (char *)malloc((size_t)len + 1);
    if (!buf) {
        wickra_compile_free(compiler);
        return 1;
    }
    wickra_compile_command(compiler, CMD, buf, (size_t)len + 1);

    printf("wickra-compile %s\n", wickra_compile_version());
    printf("output: %s\n", buf);

    free(buf);
    wickra_compile_free(compiler);
    return 0;
}

File: examples/go/compile.go
// A runnable Go example: compile a strategy spec (dry run) and print its
// deterministic manifest.
//
//   cd examples/go && go run compile.go
//
// Every language example uses the same spec and prints the same project_hash.
package main

import (
	"encoding/json"
	"fmt"

	wickra "github.com/wickra-lib/wickra-compile/bindings/go"
)

const spec = `{"strategy":{"symbol":"btcusdt","timeframe":"1h",` +
	`"indicators":{"fast":{"type":"Sma","params":[10]},"slow":{"type":"Sma","params":[30]}},` +
	`"entry":{"cross_above":["fast","slow"]},"exit":{"cross_below":["fast","slow"]},` +
	`"sizing":{"type":"fixed_qty","qty":1}},"target":{"kind":"wasm"},"crate_name":"demo"}`

func main() {
	compiler := wickra.New()
	defer compiler.Close()

	resp, err := compiler.Command(`{"cmd":"compile","dry_run":true,"spec":` + spec + `}`)
	if err != nil {
		panic(err)
	}
	var out struct {
		Manifest struct {
			CrateName   string            `json:"crate_name"`
			Files       []json.RawMessage `json:"files"`
			ProjectHash string            `json:"project_hash"`
		} `json:"manifest"`
	}
	if err := json.Unmarshal([]byte(resp), &out); err != nil {
		panic(err)
	}

	fmt.Printf("wickra-compile %s\n", wickra.Version())
	fmt.Printf("crate: %s\n", out.Manifest.CrateName)
	fmt.Printf("files: %d\n", len(out.Manifest.Files))
	fmt.Printf("project_hash: %s\n", out.Manifest.ProjectHash)
}
